### PR TITLE
Introduce LabelBox widget

### PR DESCRIPTION
**Pull request by kosme**
Fixed some wording and added a couple of instructions to make the development documentation clearer

### CODE REVIEW

**Comment by martinohanlon** [comment]
I noticed you have pushed other commits for a `LabelBox` widget to this PR.  Are you proposing this for inclusion in guizero as well as the documentation updates?

**Comment by kosme** [comment]
Yes. I had not included those on the current pull request yet since I was still working on that widget, but I think it is ready now.

**Comment by martinohanlon** [comment]
OK, I am going to change the name of the PR then. The addition of a new widget is the more "impactful" change. It would have been great to have the changes in 2 different PRs but not to worry :)

File: guizero/LabelBox.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+from tkinter import LabelFrame
+from . import utilities as utils
+from .base import ContainerTextWidget
+from .event import EventManager
+
+class LabelBox(ContainerTextWidget):
```
Comment (martinohanlon) [inline guizero/LabelBox.py:6]: Should `LabelBox` inherit from `Box`? `border`, `set_border` and `resize` methods are the same.

File: guizero/LabelBox.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        master,
+        text="",
+        layout="auto",
+        grid=None,
+        align=None,
+        visible=True,
+        enabled=None,
+        width=None,
+        height=None,
+        border=None):
```
Comment (martinohanlon) [inline guizero/LabelBox.py:19]: if `border` is None, does the LabelBox have a border by default? That seems to be the default for tkinter.LabelFrame but I havent checked.

File: guizero/LabelBox.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            The starting width of the widget. Defaults to `None` and will auto
+            size. If not `None`, both the width and height need to be set.
+
+        :param int height:
+            The starting height of the widget. Defaults to `None` and will auto
+            size. If not `None`, both the width and height need to be set.
+
+        :param int border:
+            Sets the border thickness. `0` or `False` is no border. `True` or
+            value > 1 sets a border. The default is `None`.
```
Comment (martinohanlon) [inline guizero/LabelBox.py:60]: Similar question - what is the default behaviour of LabelFrame? Does it have a border? If so, should this be 1?!

File: examples/all_widgets.py [hunk 4]
```diff
@@ -43,6 +43,9 @@
 
 # waffle = Waffle(app)
 
+# labelBox = LabelBox(app, "This is a label")
```
Comment (martinohanlon) [inline examples/all_widgets.py:46]: Is there a reason this is commented out?

File: examples/labelbox.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from guizero import App, LabelBox, Text
+app = App()
+labelbox = LabelBox(app, "Label box")
+Text(labelbox,"Text inside a LabelBox")
+
+labelbox2 = LabelBox(app)
```
Comment (martinohanlon) [inline examples/labelbox.py:6]: Not sure of the benefit of a `LabelBox` without a label. How is this be different to a `Box`?

File: guizero/LabelBox.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+from . import utilities as utils
+from .base import ContainerTextWidget
+from .event import EventManager
+
+class LabelBox(ContainerTextWidget):
+
+    def __init__(
+        self,
+        master,
+        text="",
```
Comment (martinohanlon) [inline guizero/LabelBox.py:11]: Should `text` be mandatory? What is a `LabelBox` without a label?

**Review by martinohanlon — COMMENTED** [review]
Thanks for the submission.

I have made a few comments, please take them as they are "comments" not criticism :)

I think the addition of a Box with a "title" is a good idea. 

Key points I would appreciate some opinions on:

- "Label" - isnt a term used in guizero - isnt there a different name which fits better? `TitleBox`?  @lawsie - you got any thoughts on this?
- What is the default behaviour of a `border` for `LabelFrame` - is the border shown? If so, I think the default should reflect that.
- Can `LabelBox` inherit from `Box` rather than `Container`? I think so.

**Comment by martinohanlon** [comment]
note - `docs-src/docs/widgetoverview.md` will need updating

**Comment by lawsie** [comment]
Thanks for this PR @kosme :) I would agree with all of Martin's comments. In the docs you've said a LabelBox is "an visible container which can contain other widgets with a border surrounding them, and show an optional label." but later on you say that it is invisible except for the title. I think we need to be clear on what the purpose of this new widget is, because as Martin says if it's a LabelBox without a label then that seems rather odd! A special widget for a box with just a border seems like a bit of a pointless widget though, as you could just create a Box and give it a border? `TitleBox` sounds good to me.

File: guizero/LabelBox.py (hunk 2 above)
Comment (kosme) [reply guizero/LabelBox.py:19]: The default behavior is that it has a frame around the frame's content. The setting border to anything different than 0 or None will draw a line of varying thickness around the widget AND the label effectively showing two borders. The attached image makes it clearer
labelbox-border

Therefore, I will add additional parameters to control the frame

File: examples/labelbox.py (hunk 5 above)
Comment (kosme) [reply examples/labelbox.py:6]: Indeed the results are quite similar. A Box has a default behavior of not displaying a frame and if displayed, the frame color is customizable. The default behavior of a LabelBox is displaying a frame that will be similar in color to the background color of the LabeBox.

File: examples/all_widgets.py (hunk 4 above)
Comment (kosme) [reply examples/all_widgets.py:46]: All the other widgets are commented out in this example file so I just did the same.

**Comment by kosme** [comment]
I will rename the widget as TitleBox, avoid TitleBoxes without a title, and make sure the frame size is customizable but always visible.

**Comment by martinohanlon** [comment]
I did some testing. I noticed an issue and something to consider.

Currently you cant change the *text* properties (font, text_color, text_size) of the `TitleBox`.  Fixable with a bit of thinking, probably by modifying the `ContainerTextWidget` base class.

However as `TitleBox` is a container, when you call these methods guizero will propagate font, etc to the child widgets. So if you set the title of the titlebox to be text_size=12, all the widgets in the TitleBox will also have the text_size set to 12.

I think that is "ok", but is it desirable? e.g. if you set the text_color of the `TitleBox`  would you also want the text_color of all the widgets to also be "red". My gut feel says that `TitleBox` should work in the same way as all containers. However if the user didnt want that to happen they would have to set the values individually for each widget inside the titlebox.

**Comment by martinohanlon** [comment]
A note for me...  I noticed that `ButtonGroup` inherits from a specific base class called `ContainerTextWidget`. `ContainerTextWidget` is functionally the same as `ContainerWidget` as the properties from `TextMixin` are overridden by the properties inherited from `Container`.  It looks like a bit of legacy code which can be tided up. This give the opportunity to modify `ContainerTextWidget` for `TitleBox`.

**Comment by martinohanlon** [comment]
merged into `titlebox` branch to review, test and update based on comments.

**Comment by martinohanlon** [comment]
I have made some changes to the `TitleBox` widget, mainly around simplifying the code and api and some consistency changes to bring it in line with guizero's API.

@kosme you can see the changes in this commit - https://github.com/lawsie/guizero/commit/a601026fc3ee955ed98905f5ec2e6c0cde4010da

Currently you cannot change the text properties of a TitleBox. I have done some protyping but I need to give it a bit more thought to ensure it doesnt mess up the inherit / propogate features in guizero.

**Comment by martinohanlon** [comment]
This has been merged into `dev`. I had to do some refactoring of TextMixin in order to support a widget which was a container but also supported a text properties. This is the [final commit](https://github.com/lawsie/guizero/commit/668c73db094a6442762a21e60145df2ed54b2793).

**Comment by lawsie** [comment]
Nice! :)

**Pull request merged**